server/lib/live/shared/muxing-session.ts

```diff
@@ -277,12 +281,66 @@ class MuxingSession extends EventEmitter {
     })
   }
 
+  private async parseM3U8Segments (srcDirectory: string, m3u8Path: string) {
+    const m3u8Data = await readFile(m3u8Path)
+    this.m3u8Parser = new M3u8Parser()
+    this.m3u8Parser.push(m3u8Data)
+    this.m3u8Parser.end()
+    const m3u8Parsed = this.m3u8Parser.manifest
+    const segmentPaths = m3u8Parsed.segments.map(segment => srcDirectory + '/' + segment.uri)
+    return segmentPaths
+  }
+
+  private getUnavailableM3U8Segments (m3u8SegmentPaths: string[]) {
+    const storedSegments = new Set(this.liveSegmentsInStorage)
+    const unavailableM3U8Segments = new Set(m3u8SegmentPaths)
+    for (const segment of m3u8SegmentPaths) {
+      if (storedSegments.has(segment)) {
+        unavailableM3U8Segments.delete(segment)
+      }
+    }
+    return Array.from(unavailableM3U8Segments)
+  }
+
+  private async waitForUnavailableM3U8Segments (m3u8Path: string, m3u8SegmentPaths: string[], maxRetries = 30) {
+    const retrySleepMs = 100
+    let ackSleepMs = 0
+    while (maxRetries) {
```

Comment:
```suggestion
  private async waitForUnavailableM3U8Segments (m3u8Path: string, m3u8SegmentPaths: string[], maxRetries = 40) {
    const retrySleepMs = 250
    let ackSleepMs = 0
    while (maxRetries) {
```

I think that the total timeout of 3 seconds here is a bit too short. When test running this on a dev server for 30 minutes, I got the "max retries exceeded" warning a bit over 2200 times. When streaming a live feed at around 20,000 kbit/s and transcoding into 1080p, 720p, 480p, 240p, and 144p, these warnings appear about every 4 seconds.

Illustration from debug logs:
```plain
$ grep "retries left: 0" 230424_stdout-debug.log
[pt-dev.mabl.online:443] 2023-04-24 16:19:35.245 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/0.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 3000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:19:35.436 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/1.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 3000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:19:35.437 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/2.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 3000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:19:35.439 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/4.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 3000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:19:35.445 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/3.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 3000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:19:39.303 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/0.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 3000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:19:39.498 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/1.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 3000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:19:39.499 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/2.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 3000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:19:39.501 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/3.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 3000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:19:39.502 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/4.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 3000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:19:43.453 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/0.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 3000 ms, retries left: 0) {
/.../
```

After adjusting the total timeout to 5 seconds, 20 x 250 ms, I get only 6 "max retries exceeded" after 10 minutes:
```plain
grep "retries left: 0" 230424_stdout-debug-5s.log && date
[pt-dev.mabl.online:443] 2023-04-24 16:30:20.374 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/0.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 5000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:30:20.605 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/3.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 5000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:30:20.876 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/4.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 5000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:33:52.632 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/3.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 5000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:33:52.800 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/1.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 5000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:34:28.276 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/0.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 5000 ms, retries left: 0) {
[pt-dev.mabl.online:443] 2023-04-24 16:38:36.402 warn: Max retries exceeded waiting for /home/ar9708/git/PeerTube/storage/streaming-playlists/hls/8801a1ac-6810-4c5a-9315-c8f70b69c73c/0.m3u8 video segment(s) in storage, giving up (segments: 1, wait time: 5000 ms, retries left: 0) {
Mon Apr 24 16:41:23 UTC 2023
```

I will try to run a long-running test on 10 second total timeout [with this video](https://pt-dev.mabl.online/w/hN6c8crzhsTciZK9yDJhvb) and see if it's performing without any error. Have you had any time to look into this, @Chocobozzz ?

**Note:** I have about 500 Mbps in fiber uplink speed, so I don't think my connection is causing the problem, but rather some inherent latency in S3 infra/sync.

Reply:
Update from the longer test run, @Chocobozzz & @mblomdahl – after running for 90 minutes with 10 seconds total timeout, 0 "max retries exceeded" occurred. Here is a chart I put together with the variance in delays:
![230424_Time in milliseconds until all m3u8 segments are in S3](https://user-images.githubusercontent.com/123388211/234088669-b6541186-06be-4d24-ba3e-01ba0e52a437.png)
CSV with source data: [230424_Time in milliseconds until all m3u8 segments are in S3.csv](https://github.com/Chocobozzz/PeerTube/files/11313672/230424_Time.in.milliseconds.until.all.m3u8.segments.are.in.S3.csv)

For this being a server with higher network latency or shaky bandwidth, my guesstimation is that 10 seconds is fairly good for other production servers too. If a server can't publish all live stream segments within this timeout, then it is probably advisable to consider streaming in 480p only (for example) instead of full 1080p.